### ISSUE TITLE
Why are certain models with a higher WER (on the eval set) performing better or as good as  models with a lower WER – when tested on the test set?

### ISSUE THREAD

**Issue by drishtishrrma** [huggingface/transformers#15664]
Hello @patrickvonplaten and @anton-l  !

I had two weird observations while evaluating my models, one of which showed that the model with the higher WER clearly outperformed the one with the lower WER (So strange!). In another scenario, when evaluated on the test set, the model with higher WER (on the eval set) fared "almost" as good as the model with low WER. I'm curious as to what the likely causes of these observations are. 

I compared two models (Model A against model B) in each case. So, in a nutshell, I’ll be mentioning two models in each case for the sake of comparison.

**Outline of the issue:**

Case 1: Model A and Model B — trained for Guarani language (gn) —- CV8 dataset — model with higher  WER on eval set clearly outperformed the model with lower WER, when tested on the test set.

Case 2: Different Model A and Model B — trained for Assamese language (as) —- CV8 dataset —- model with higher WER performed nearly as good as model with low WER, when tested on the test set.

**Detailed Description:** 

**Case 1:** 
Models A and B were trained for the Guarani (**gn**) – Common Voice 8 dataset. 

**It should be noted that the library versions of Datasets and Transformers utilized while training both models were different.** **However, those models were tested in literally the same environment setting ie. the same version of Datasets and Transformers**. And it turns out that when tested on the test set, the model with a higher WER on the eval set outperformed the model with a lower WER.

Case 1: Model A: [https://huggingface.co/lgris/wav2vec2-xls-r-300m-gn-cv8](url)

Case 1: Model B: [https://huggingface.co/DrishtiSharma/wav2vec2-large-xls-r-300m-gn-k1](url)

**Comparison in Tabular form**: 
![image](https://user-images.githubusercontent.com/82262606/154126093-d6bd46c7-adfe-496c-a791-f2d597a8c7c2.png)

**Case 2**: In this case, different model A and model B were trained for Assamese language (as) – Common Voice 8 dataset. 

It's worth noting that **the Datasets version used in both the cases were same ie. Datasets 1.18.3. However the Transformers version was different.** And it turns out that when tested on the test set, the model with a higher WER performed nearly as good as the model with a lower WER. 

Case 2: Model A: [https://huggingface.co/DrishtiSharma/wav2vec2-large-xls-r-300m-as-g1 ](url)

Case 2: Model B: [https://huggingface.co/infinitejoy/wav2vec2-large-xls-r-300m-assamese-cv8 ](url)

**Comparison in Tabular form**: 

![image](https://user-images.githubusercontent.com/82262606/154129659-b9a49041-3366-4b4d-a52f-2670a971b914.png)

I'm curious as to what could be the underlying causes of such observations. I'd be glad if I get your inputs in this regard.
Currently, I'm a super novice in the ASR field. I literally calculated the WER for the very first time in this sprint itself, so please pardon me if I'm witnessing these observations due to some mistake on my part.

Thanks in advance...

**Comment by LysandreJik** [comment]
Hello, thanks for opening an issue! We try to keep the github issues for bugs/feature requests.
Could you ask your question on the [forum](https://discuss.huggingface.co) instead?

Thanks!

**Comment by drishtishrrma** [comment]
Hi @LysandreJik!

I opened this issue since it was requested.. A bug in Datasets library was reported a few days back due to which we gotta new version now. I was actually referring to that bug when I was comparing the Datasets version utilised in different cases This issue was produced in an attempt to detect the underlying causes of the models' anomalous behaviour.

But I'm guessing I didn't elucidate properly. Perhaps, I'd have used the default format to explain the issue, however I  thought it's easier this way since there were too many variables.. :/

**Comment by drishtishrrma** [comment]
@LysandreJik :

Anyway, if you still believe that the aforementioned issue/post is forum- worthy, I have no objection in taking it down and publishing it on forums.
Muti-channel communication is turning chaotic for me as well. And I've a strong intuition that I've almost found the underlying causes of anomalous behaviour of certain models, albeit I'm still refining my points and I'm almost there... I might get my points validated on the forum itself! No worries! 

Lastly, I apologise if the above issue is too naive/inappropriate for the platform. I've transitioned from the Patent Research and Analytics domain, and haven't utilized GitHub a lot more. However, I hope to do better in future. 

Thanks!

**Comment by LysandreJik** [comment]
Ah, sorry, if it was asked by a member of the team then please ignore my message above. Thank you for doing the effort of opening an issue, even if I mentioned the forum would be a better place we definitely appreciate you trying to solve the problem. The issue is neither too naive nor inappropriate, it was just more likely to get an answer on the forum :)

**Comment by anton-l** [comment]
Hey @drishtishrrma! I've re-evaluated the models over the past week and haven't found any significant model or config-related differences that would explain these weird coincidences :slightly_smiling_face: 
However, note that the Guarani and Assamese languages have **very** little data in the Common Voice dataset (40m and 1h of validated data respectively). This means that the validation and testing metrics can be highly uncorrelated (e.g. there are just 293 train, 159 val and 93 test sentences for Guarani). In these low-resource conditions even changing the random seed can drastically change the final model.
Hopefully that makes sense, and thank you for the evaluation report! Let me know if you have any questions :slightly_smiling_face:

**Comment by drishtishrrma** [comment]
@LysandreJik : Yeah, I got your POV. Thanks for your advice, I appreciate it...

**Comment by drishtishrrma** [comment]
@anton-l:

Small batch_size achieves better generalization, this claim needs to be tested a little. Recently I came across a couple tweets touting the benefits of larger batch_size. I'll be putting it to test soon. 

**To simplify things and avoid making claims I haven't personally tested, I'd like to generalize the statement I stated above as follows:**
Because different batch size (s) were utilized, it's likely that one of the models generalised better for the limited data available.

In short, let say,  if a model generalized on the data which closely matches the environment in which we're gonna be testing, then obviously it'll perform better. And on the contrary, it might not perform well if subjected to envio for which it didn't generalize well during the training. 

This effect gets more pronounced in the low-resource scenario. And it's very likely that if the aforementioned models are exposed to a different kind of environment, the outcome could be the polar opposite of what we've seen so far.
